course/src/matrix3.cpp

```diff
@@ -0,0 +1,172 @@
+
+namespace ekumen {
+namespace math {
+
+const Matrix3 Matrix3::kIdentity{Matrix3(1., 0., 0., 0., 1., 0., 0., 0., 1.)};
+const Matrix3 Matrix3::kOnes{Matrix3(1., 1., 1., 1., 1., 1., 1., 1., 1.)};
+const Matrix3 Matrix3::kZero{Matrix3(0., 0., 0., 0., 0., 0., 0., 0., 0.)};
+
+Matrix3::Matrix3() : a_{0., 0., 0.}, b_{0., 0., 0.}, c_{0., 0., 0.} {}
+
+Matrix3::Matrix3(const double a1, const double a2, const double a3,
+                 const double b1, const double b2, const double b3,
+                 const double c1, const double c2, const double c3)
+    : a_{a1, a2, a3}, b_{b1, b2, b3}, c_{c1, c2, c3} {}
+
+bool Matrix3::operator==(const Matrix3& matrix) const {
+  return a_ == matrix[0] && b_ == matrix[1] && c_ == matrix[2];
+}
+
+bool Matrix3::operator!=(const Matrix3& matrix) const {
+  return !(*this == matrix);
+}
+
+Vector3 Matrix3::operator[](const int index) const {
+  switch (index) {
+    case 0:
+      return a_;
+    case 1:
+      return b_;
+    case 2:
+      return c_;
+    default:
+      throw std::out_of_range("Matrix3 has only 3 elements");
+  }
+}
+
+Vector3& Matrix3::operator[](const int index) {
+  switch (index) {
+    case 0:
+      return a_;
+    case 1:
+      return b_;
+    case 2:
+      return c_;
+    default:
+      throw std::out_of_range("Matrix3 has only 3 elements");
+  }
+}
+
+std::ostream& operator<<(std::ostream& os, const Matrix3& matrix) {
+  os << "[[" << matrix[0][0] << ", " << matrix[0][1] << ", " << matrix[0][2]
+     << "], [" << matrix[1][0] << ", " << matrix[1][1] << ", " << matrix[1][2]
+     << "], [" << matrix[2][0] << ", " << matrix[2][1] << ", " << matrix[2][2]
+     << "]]";
+  return os;
+}
+
+Matrix3& Matrix3::operator+=(const Matrix3& matrix) {
+  a_ += matrix[0];
+  b_ += matrix[1];
+  c_ += matrix[2];
+  return *this;
+}
+
+Matrix3& Matrix3::operator-=(const Matrix3& matrix) {
+  a_ -= matrix[0];
+  b_ -= matrix[1];
+  c_ -= matrix[2];
+  return *this;
+}
+
+Matrix3& Matrix3::operator*=(const Matrix3& matrix) {
+  a_ *= matrix[0];
+  b_ *= matrix[1];
+  c_ *= matrix[2];
+  return *this;
+}
+
+Matrix3& Matrix3::operator*=(double scalar) {
+  a_ *= scalar;
+  b_ *= scalar;
+  c_ *= scalar;
+  return *this;
+}
+
+Matrix3& Matrix3::operator/=(const Matrix3& matrix) {
+  a_ /= matrix[0];
+  b_ /= matrix[1];
+  c_ /= matrix[2];
+  return *this;
+}
+
+Matrix3& Matrix3::operator/=(double scalar) {
+  a_ /= scalar;
+  b_ /= scalar;
+  c_ /= scalar;
+  return *this;
+}
+
+Matrix3 Matrix3::operator+(const Matrix3& matrix) const {
+  Matrix3 aux{*this};
+  aux += matrix;
+  return aux;
+}
+
+Matrix3 Matrix3::operator-(const Matrix3& matrix) const {
+  Matrix3 aux{*this};
+  aux -= matrix;
+  return aux;
+}
+
+Matrix3 Matrix3::operator*(const Matrix3& matrix) const {
+  Matrix3 aux{*this};
+  aux *= matrix;
+  return aux;
+}
+
+Vector3 Matrix3::operator*(const Vector3& vector) const {
+  Vector3 aux;
+  aux[0] = a_.dot(vector);
+  aux[1] = b_.dot(vector);
+  aux[2] = c_.dot(vector);
+  return aux;
```

Comment:
nit: You can just simply do:

```suggestion
  return {a_.dot(vector), b_.dot(vector), c_.dot(vector)};
```